bitswap/provider.go

```diff
@@ -252,17 +263,20 @@ func (provider *Provider) handleRequest(
 		Priority: int(entry.Priority),
 		Work:     1,
 	})
+	provider.retrievalQueueSignalChan <- struct{}{}
 
 	return nil
 }
 
-func (provider *Provider) handleResponses() {
-	ctx := context.Background()
-
-	for {
+func (provider *Provider) handleResponses(ctx context.Context) {
+	for ctx.Err() == nil {
 		peerID, tasks, _ := provider.responseQueue.PopTasks(targetMessageSize)
 		if len(tasks) == 0 {
-			time.Sleep(time.Millisecond * 250)
+			select {
+			case <-ctx.Done():
+			case <-time.After(time.Millisecond * 250):
+			case <-provider.responseQueueSignalChan:
```

Comment:
when len(tasks) != 0, you had better still optionally drain the signal chan, i.e

```golang
if len(tasks == 0) {
   ///...
   continue
}

select {
case <-provider.responseQueueSignalChan:
default:
}
///...

Reply:
I wouldn't be surprised if you're getting stuck cause of this.